### PR TITLE
pallet-asset: Fix transfer of a large amount of an asset

### PR DESCRIPTION
**Pull request by bkchr**
Before this pr transferring a large amount of an asset would check that transferring the asset would
not overflow the supply of the asset. However, it doesn't make sense to check for asset supply
overflow when we just transfer from one account to another account and don't increase the supply in
any way. It also required to extend the `can_deposit` method of `fungible` and `fungibles` with a
`mint` parameter. If this parameter is set to `true`, it means we want to mint the amount of an
asset before transferring it into an account. For `can_withdraw` we don't need to add an extra
parameter, because withdrawing should never be able to underflow the supply. If that would happen,
it would mean that somewhere the supply wasn't increased while increasing the balance of an account.

### CODE REVIEW

File: frame/assets/src/functions.rs [hunk 1]
```diff
@@ -535,8 +543,8 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 		}
 
 		// Figure out the debit and credit, together with side-effects.
-		let debit = Self::prep_debit(id, &source, amount, f.into())?;
-		let (credit, maybe_burn) = Self::prep_credit(id, &dest, amount, debit, f.burn_dust)?;
+		let debit = dbg!(Self::prep_debit(id, &source, amount, f.into()))?;
```
Comment (xlc) [inline frame/assets/src/functions.rs:546]: do you intent to have the `dbg` here?

Comment (bkchr) [reply frame/assets/src/functions.rs:546]: Ohh 😅 no

Comment (bkchr) [inline frame/assets/src/functions.rs:546]:
```suggestion
		let debit = Self::prep_debit(id, &source, amount, f.into())?;
```

File: frame/assets/src/functions.rs [hunk 2]
```diff
@@ -536,7 +544,7 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 
 		// Figure out the debit and credit, together with side-effects.
 		let debit = Self::prep_debit(id, &source, amount, f.into())?;
-		let (credit, maybe_burn) = Self::prep_credit(id, &dest, amount, debit, f.burn_dust)?;
+		let (credit, maybe_burn) = dbg!(Self::prep_credit(id, &dest, amount, debit, f.burn_dust))?;
```
Comment (bkchr) [inline frame/assets/src/functions.rs:547]:
```suggestion
		let (credit, maybe_burn) = Self::prep_credit(id, &dest, amount, debit, f.burn_dust)?;
```

File: frame/assets/src/functions.rs [hunk 3]
```diff
@@ -103,16 +103,24 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 		Remove
 	}
 
+	/// Returns `true` when the balance of `account` can be increased by `amount`.
+	///
+	/// - `id`: The id of the asset that should be increased.
+	/// - `who`: The account of which the balance should be increased.
+	/// - `amount`: The amount by which the balance should be increased.
+	/// - `will_increase_supply`: Will the supply of the asset be increased by `amount` at the same
```
Comment (shawntabrizi) [inline frame/assets/src/functions.rs:111]:
```suggestion
	/// - `increase_supply`: Will the supply of the asset be increased by `amount` at the same
```

or update the variable below

**Comment by shawntabrizi** [comment]
fortunately, all of this complexity can probably be re-written and removed once transactional storage by default is implemented

**Comment by bkchr** [comment]
> fortunately, all of this complexity can probably be re-written and removed once transactional storage by default is implemented

Yep :)

**Review by shawntabrizi — APPROVED** [review]

File: frame/assets/src/functions.rs [hunk 4]
```diff
@@ -103,16 +103,24 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 		Remove
 	}
 
+	/// Returns `true` when the balance of `account` can be increased by `amount`.
+	///
+	/// - `id`: The id of the asset that should be increased.
+	/// - `who`: The account of which the balance should be increased.
+	/// - `amount`: The amount by which the balance should be increased.
+	/// - `increase_supply`: Will the supply of the asset be increased by `amount` at the same
+	///   time as crediting the `account`.
```
Comment (kianenigma) [inline frame/assets/src/functions.rs:112]:
```suggestion
	///   time as crediting `amount` to `who`.
```

**Review by kianenigma — APPROVED** [review]

**Pull request merged**